api/genericresource/resource_management.go

```diff
@@ -0,0 +1,62 @@
+package genericresource
+
+import (
+	"github.com/docker/swarmkit/api"
+)
+
+// Claim assigns GenericResources to the task by taking them from the node's GenericResource
+// and storing them in the taken store
+func Claim(nodeRes *api.GenericResource, taskRes *api.GenericResource, taken *api.GenericResource) error {
+	t, err := getGenericResourceDiscrete(taskRes)
+	if err != nil {
+		return err
+	}
+
+	switch nr := nodeRes.Resource.(type) {
+	case *api.GenericResource_Discrete:
+		if nr.Discrete < t.Discrete {
+			return NewError("cannot provide enough of scalar resource requested %d, available %d",
+				t.Discrete, nr.Discrete)
+		}
+		nr.Discrete -= t.Discrete
+		taken.Resource = NewDiscrete(t.Discrete)
+	case *api.GenericResource_Set:
+		i := uint64(0)
+		res := NewSet()
+
+		for k := range nr.Set.Val {
+			if i >= t.Discrete {
+				break
+			}
+
+			i++
+
+			res.Set.Val[k] = nil
+			if _, ok := nr.Set.Val[k]; !ok {
+				return NewError("missing key %s from node resource", k)
+			}
+			delete(nr.Set.Val, k)
+		}
+
+		taken.Resource = res
+	}
+
+	return nil
+}
+
+// Reclaim adds the resources taken by the task to the taken store
+func Reclaim(nodeRes *api.GenericResource, taskRes *api.GenericResource, taken *api.GenericResource) error {
+
+	switch tr := taken.Resource.(type) {
+	case *api.GenericResource_Discrete:
+		nodeRes.Resource.(*api.GenericResource_Discrete).Discrete += tr.Discrete
+	case *api.GenericResource_Set:
+		set := nodeRes.Resource.(*api.GenericResource_Set).Set.Val
+
+		for k := range tr.Set.Val {
+			set[k] = nil
+		}
```

Comment:
`Reclaim` is called by `removeTask`. My thinking was...

I have node N with `{apple,orange,grapefruit}`. I schedule a task that needs a generic resource and assign it `apple`. But later on, the node gets updated to remove `apple`. This means the constraint enforcer will kill that task and cause it to be restarted. When the scheduler sees that the task has been killed, it will call `removeTask`, which will return the resources used by the task to the available set. But since the node doesn't have `apple` anymore, it *shouldn't* be returned to the available set.